main.go

```diff
@@ -350,3 +389,137 @@ func webhooks() []webhookwrapper.WebhookRegistrator {
+//
+// controllerWebhookService is the service name used to access the webhook server
+// available by the controller. This is necessary to properly generate the
+// certificate
+//
+// deploymentsNamespace is the namespace where the controller is deployed
+//
+// validatingWebhooks are the ValidatingWebhookConfiguration names which should
+// be updated when the root CA is initialized
+//
+// mutatingWebhooks are the MutatingWebhookConfiguration names which should
+// be updated when the root CA is initialized
+//
+// developmentMode is a flag to tell if the controller is running on
+// development mode or not
+func setupCA(ctx context.Context, clusterClient client.Client, controllerWebhookService string, deploymentsNamespace string, validatingWebhooks []string, mutatingWebhooks []string, developmentMode bool) ([]byte, []byte, bool, error) {
+	caSecret, caInitialized, err := admission.FetchOrInitializeCARootSecret(ctx, clusterClient, deploymentsNamespace, admissionregistration.GenerateCA, admissionregistration.PemEncodeCertificate)
+	if err != nil {
+		return []byte{}, []byte{}, false, fmt.Errorf("failed to fetch or initialize root CA certificate secret: %s", err.Error())
+	}
+	if caInitialized {
+		if err = admission.ReconcileSecret(ctx, clusterClient, caSecret); err != nil {
+			return []byte{}, []byte{}, false, fmt.Errorf("failed to reconcile root CA certificate secret: %s", err.Error())
+		}
+	}
+
+	certificate, privateKey, err := admission.ExtractCertificateData(caSecret)
+	if err != nil {
+		return []byte{}, []byte{}, false, fmt.Errorf("failed to initialize root CA certificate")
+	}
+
+	controllerSecret, initialized, err := admission.FetchOrInitializeCertificate(ctx, clusterClient, controllerWebhookService, deploymentsNamespace, constants.ControllerCertificateSecretName, caSecret, admissionregistration.GenerateCert)
+	if err != nil {
+		return []byte{}, []byte{}, false, fmt.Errorf("failed to fetch or init the controller certificate: %s ", err.Error())
+	}
+	if initialized {
+		if err = admission.ReconcileSecret(ctx, clusterClient, controllerSecret); err != nil {
+			return []byte{}, []byte{}, false, fmt.Errorf("cannot reconcile secret %s: %s", controllerSecret.Name, err.Error())
+		}
+	}
+
+	if !developmentMode {
+		if err := setupCAWebhooksInDevelopmentMode(ctx, clusterClient, certificate, validatingWebhooks, mutatingWebhooks); err != nil {
+			return []byte{}, []byte{}, false, fmt.Errorf("failed to configure webhooks in development mode: %s", err.Error())
+		}
+	}
+	return certificate, privateKey, (caInitialized || initialized), nil
+}
+
+func setupCAWebhooksInDevelopmentMode(ctx context.Context, clusterClient client.Client, certificate []byte, validatingWebhooks []string, mutatingWebhooks []string) error {
+	// If the controller is NOT running in development mode, the
+	// webhooks are created by Helm chart and missing the
+	// `caBundle` field with the root CA certificate. Therefore,
+	// the controller needs to patch these resources
```

Comment:
That means that when the controller is deployed via helm our helm charts take care of creating the webhook configurations? Can't we have the controller always create them? I mean, we already have the code somewhere to make the development environment work... This would make everything more consistent

Reply:
Moreover, the name of this function is confusing, it should be `setupCAWebhooksInProductionMode`, but again, I hope we can get rid of this code and have the webhook configurations managed by the controller in any kind of deployment environment